### PR TITLE
Preparatory work for the double_real external project

### PR DESCRIPTION
**Pull request by jfalcou**

### CODE REVIEW

File: include/eve/module/core/pedantic/impl/is_not_flint.hpp [hunk 1]
```diff
@@ -24,8 +24,9 @@ is_not_flint_(EVE_SUPPORTS(cpu_), pedantic_type const&, T const& a) noexcept
 {
   if constexpr( has_native_abi_v<T> )
   {
-    if constexpr( integral_value<T> ) return true_(eve::as<T>());
-    if constexpr( floating_value<T> ) return is_nez(frac(a)) || (a > eve::maxflint(eve::as<T>()));
+    using u_t =  underlying_type_t<T>;
+    if constexpr( integral_value<u_t> ) return true_(eve::as<T>());
+    if constexpr( floating_value<u_t> ) return is_nez(frac(a)) || (a > eve::maxflint(eve::as<T>()));
```
Comment (jfalcou) [inline include/eve/module/core/pedantic/impl/is_not_flint.hpp:29]: else if

File: include/eve/module/core/numeric/impl/fma.hpp [hunk 2]
```diff
@@ -48,14 +45,23 @@ fma_(EVE_SUPPORTS(cpu_), numeric_type const&, T const& a, T const& b, T const& c
...
-    return pedantic(ldexp)(s + (rp + rs), -e0);
+    if constexpr(scalar_value<T>)
+    {
+      return std::fma(a, b, c);
+    }
+    else
+    {
+      auto stdfma = [](auto sa, auto sb, auto sc){return std::fma(sa, sb, sc); };
+      return map(stdfma, a, b, c);
+//       auto amax    = maxmag(a, b);
```
Comment (jfalcou) [inline include/eve/module/core/numeric/impl/fma.hpp:56]: remove don't comment

File: include/eve/module/core/numeric/impl/maxabs.hpp [hunk 3]
```diff
@@ -34,10 +34,24 @@ maxabs_(EVE_SUPPORTS(cpu_),
...
 {
   return eve::numeric(max)(eve::abs(a), eve::abs(b));
 }
+
+//================================================================================================
+// N parameters
+//================================================================================================
+template<ordered_value T0, ordered_value T1, ordered_value... Ts>
+auto
+maxabs_(EVE_SUPPORTS(cpu_), numeric_type const&, T0 a0, T1 a1, Ts... args)
```
Comment (jfalcou) [inline include/eve/module/core/numeric/impl/maxabs.hpp:49]: miss noexcept

File: include/eve/module/core/numeric/impl/negmaxabs.hpp [hunk 4]
```diff
@@ -42,7 +42,7 @@ negmaxabs_(EVE_SUPPORTS(cpu_), numeric_type const&, T const& a, T const& b) noex
 template<value T0, value T1, value... Ts>
 auto
 negmaxabs_(EVE_SUPPORTS(cpu_), numeric_type const&, T0 a0, T1 a1, Ts... args)
```
Comment (jfalcou) [inline include/eve/module/core/numeric/impl/negmaxabs.hpp:44]: miss noexcept

File: include/eve/module/core/pedantic/impl/fma.hpp [hunk 5]
```diff
@@ -46,9 +44,18 @@ fma_(EVE_SUPPORTS(cpu_), pedantic_type const&, T const& a, T const& b, T const&
...
-    return s + (rp + rs);
+    if constexpr(scalar_value<T>)
+    {
+      return std::fma(a, b, c);
+    }
+    else
+    {
+      auto stdfma = [](auto sa, auto sb, auto sc){return std::fma(sa, sb, sc); };
+      return map(stdfma, a, b, c);
+//       auto [p, rp] = two_prod(a, b);
```
Comment (jfalcou) [inline include/eve/module/core/pedantic/impl/fma.hpp:55]: remove don't comment

File: include/eve/module/core/pedantic/impl/is_flint.hpp [hunk 6]
```diff
@@ -23,8 +23,9 @@ is_flint_(EVE_SUPPORTS(cpu_), pedantic_type const&, T const& a) noexcept
 {
   if constexpr( has_native_abi_v<T> )
   {
-    if constexpr( integral_value<T> ) return false_(eve::as<T>());
-    if constexpr( floating_value<T> ) return is_eqz(frac(a)) && (a <= eve::maxflint(eve::as<T>()));
+    using u_t =  underlying_type_t<T>;
+    if constexpr( integral_value<u_t> ) return false_(eve::as<T>());
+    if constexpr( floating_value<u_t> ) return is_eqz(frac(a)) && (a <= eve::maxflint(eve::as<T>()));
```
Comment (jfalcou) [inline include/eve/module/core/pedantic/impl/is_flint.hpp:28]: else if

File: include/eve/module/core/regular/binarize.hpp [hunk 7]
```diff
@@ -53,6 +53,13 @@ namespace eve
 //!  @godbolt{doc/core/regular/binarize.cpp}
 //! @}
 //================================================================================================
+// namespace tag
```
Comment (jfalcou) [inline include/eve/module/core/regular/binarize.hpp:56]: remove don't comment

File: include/eve/module/core/regular/impl/three_fma.hpp [hunk 8]
```diff
@@ -7,30 +7,30 @@
...
-#include <eve/concept/value.hpp>
-#include <eve/detail/apply_over.hpp>
-#include <eve/detail/implementation.hpp>
-#include <eve/detail/kumi.hpp>
-#include <eve/module/core/regular/if_else.hpp>
-#include <eve/module/core/regular/is_not_finite.hpp>
-#include <eve/module/core/regular/two_add.hpp>
-#include <eve/module/core/regular/two_prod.hpp>
-#include <eve/arch/platform.hpp>
+// #include <eve/concept/value.hpp>
```
Comment (jfalcou) [inline include/eve/module/core/regular/impl/three_fma.hpp:10]: remove the file no ?

File: include/eve/module/core/regular/impl/rec.hpp [hunk 9]
```diff
@@ -32,7 +32,10 @@ rec_(EVE_SUPPORTS(cpu_), T a) noexcept
 {
   if constexpr( has_native_abi_v<T> )
   {
-    if constexpr( floating_value<T> ) { return T {1} / a; }
+    if constexpr( floating_value<T> )
```
Comment (jfalcou) [inline include/eve/module/core/regular/impl/rec.hpp:35]: this is noise

File: include/eve/module/core/regular/impl/two_prod.hpp [hunk 10]
```diff
@@ -15,20 +15,25 @@
...
 {
   if constexpr( has_native_abi_v<T> )
   {
-    auto [a1, a2] = two_split(a);
-    auto [b1, b2] = two_split(b);
-    T r0          = a * b;
-    T r1          = a2 * b2 - (((r0 - a1 * b1) - a2 * b1) - a1 * b2);
-    if constexpr( eve::platform::supports_invalids ) r1 = if_else(is_not_finite(r0), eve::zero, r1);
+
+//     auto [a1, a2] = two_split(a);
```
Comment (jfalcou) [inline include/eve/module/core/regular/impl/two_prod.hpp:29]: remove don't comment

File: include/eve/module/core/regular/quick_two_add.hpp [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+//!
+//!  @godbolt{doc/core/regular/quick_two_add.cpp}
+//!
+//! @}
+//================================================================================================
+namespace tag
+{
+  struct quick_two_add_;
+}
+template<> struct supports_conditional<tag::quick_two_add_> : std::false_type
```
Comment (jfalcou) [inline include/eve/module/core/regular/quick_two_add.hpp:62]: did you wrote the conditionnal cases ?

File: include/eve/module/core/regular/quick_two_sub.hpp [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+//!
+//!  @godbolt{doc/core/regular/quick_two_sub.cpp}
+//!
+//! @}
+//================================================================================================
+namespace tag
+{
+  struct quick_two_sub_;
+}
+template<> struct supports_conditional<tag::quick_two_sub_> : std::false_type
```
Comment (jfalcou) [inline include/eve/module/core/regular/quick_two_sub.hpp:63]: did you wrote the conditionnal cases ?

File: include/eve/module/core/regular/two_sqr.hpp [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+//!
+//!  @godbolt{doc/core/regular/two_sqr.cpp}
+//!
+//! @}
+//================================================================================================
+namespace tag
+{
+  struct two_sqr_;
+}
+template<> struct supports_conditional<tag::two_sqr_> : std::false_type
```
Comment (jfalcou) [inline include/eve/module/core/regular/two_sqr.hpp:61]: did you wrote the conditionnal cases ?

File: include/eve/module/core/regular/two_sub.hpp [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+//!
+//!  @godbolt{doc/core/regular/two_sub.cpp}
+//!
+//! @}
+//================================================================================================
+namespace tag
+{
+  struct two_sub_;
+}
+template<> struct supports_conditional<tag::two_sub_> : std::false_type
```
Comment (jfalcou) [inline include/eve/module/core/regular/two_sub.hpp:61]: did you wrote the conditionnal cases ?

File: include/eve/module/core/saturated/impl/negmaxabs.hpp [hunk 15]
```diff
@@ -42,10 +42,10 @@ negmaxabs_(EVE_SUPPORTS(cpu_),
 //================================================================================================
 template<ordered_value T0, ordered_value T1, ordered_value... Ts>
 auto
-negmaxabs_(EVE_SUPPORTS(cpu_), saturated_type const&, T0 a0, T1 a1, Ts... args) noexcept
+negmaxabs_(EVE_SUPPORTS(cpu_), saturated_type const&, T0 a0, T1 a1, Ts... args)
```
Comment (jfalcou) [inline include/eve/module/core/saturated/impl/negmaxabs.hpp:45]: put noexcept back

File: include/eve/module/core/regular/rec.hpp [hunk 16]
```diff
@@ -70,6 +70,14 @@ namespace eve
 //!     very poor accuracy in return. Otherwise it uses the non decorated call.
 //! @}
 //================================================================================================
+namespace tag
+{
+  struct rec_;
+}
+
+template<> struct supports_optimized_conversion<tag::rec_> : std::true_type
```
Comment (jfalcou) [inline include/eve/module/core/regular/rec.hpp:78]: Why this change ?

File: include/eve/module/math/constant/quarter.hpp [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+    using t_t = element_type_t<T>;
+    if constexpr( std::is_same_v<t_t, float> ) return T(0x1.0p-2);
+    else if constexpr( std::is_same_v<t_t, double> ) return T(0x1.0p-2);
+  }
+
+  template<floating_ordered_value T, typename D>
+  EVE_FORCEINLINE constexpr auto quarter_(EVE_SUPPORTS(cpu_), D const&, as<T> const&) noexcept
+      requires(is_one_of<D>(types<upward_type, downward_type> {}))
+  {
+    using t_t = element_type_t<T>;
```
Comment (jfalcou) [inline include/eve/module/math/constant/quarter.hpp:65]: just call the other one, this is needless duplication

File: include/eve/module/math/regular/impl/cosh.hpp [hunk 18]
```diff
@@ -31,26 +31,27 @@ cosh_(EVE_SUPPORTS(cpu_), T a0) noexcept
   }
   if constexpr( has_native_abi_v<T> )
   {
-    auto ovflimitmln2 =
-        Ieee_constant<T, 0X42AF5DC0U, 0X408628B76E3A7B61LL>(); // 87.68310404,
-                                                               // 709.08956571282405469058276787854
+    T ovflimitmln2 = maxlog(as(a0))-log_2(as(a0));
```
Comment (jfalcou) [inline include/eve/module/math/regular/impl/cosh.hpp:34]: remove comments

File: test/unit/module/math/agd.cpp [hunk 19]
```diff
@@ -39,8 +39,8 @@ TTS_CASE_WITH("Check behavior of agd on wide",
   using v_t = eve::element_type_t<T>;
   using eve::agd;
   using eve::sinh;
-  TTS_ULP_EQUAL(agd(a0), map([](auto e) -> v_t { return std::atanh(std::sin(e)); }, a0), 4);
-  TTS_ULP_EQUAL(agd(a1), map([](auto e) -> v_t { return std::atanh(std::sin(e)); }, a1), 4);
+  TTS_ULP_EQUAL(agd(a0), map([](auto e) -> v_t { return std::atanh(std::sin(e)); }, a0), 2.0e+4);//TODO
```
Comment (jfalcou) [inline test/unit/module/math/agd.cpp:42]: then do it :D

File: test/unit/module/math/cosh.cpp [hunk 20]
```diff
@@ -29,10 +29,11 @@ TTS_CASE_TPL("Check return types of cosh", eve::test::simd::ieee_reals)
 //==================================================================================================
 auto maxi = []<typename T>(eve::as<T> const&)
 {
-  using v_t = eve::element_type_t<T>;
-  v_t ovl   = eve::Ieee_constant<v_t, 0x42B0C0A4U, 0x40862E42FEFA39EFULL>(); // 88.376251220703125f,
-                                                                             // 709.782712893384
-  return T(ovl);
+ return eve::maxlog(eve::as<T>())-eve::log_2(eve::as<T>());
+//  using v_t = eve::element_type_t<T>;
```
Comment (jfalcou) [inline test/unit/module/math/cosh.cpp:33]: remove don't comment

File: include/eve/module/core/pedantic/impl/is_not_flint.hpp (hunk 1 above)
Comment (jtlap) [reply include/eve/module/core/pedantic/impl/is_not_flint.hpp:29]: done

File: include/eve/module/core/numeric/impl/fma.hpp (hunk 2 above)
Comment (jtlap) [reply include/eve/module/core/numeric/impl/fma.hpp:56]: done

File: include/eve/module/core/numeric/impl/maxabs.hpp (hunk 3 above)
Comment (jtlap) [reply include/eve/module/core/numeric/impl/maxabs.hpp:49]: done

File: include/eve/module/core/numeric/impl/negmaxabs.hpp (hunk 4 above)
Comment (jtlap) [reply include/eve/module/core/numeric/impl/negmaxabs.hpp:44]: done

Comment (jtlap) [reply include/eve/module/core/numeric/impl/negmaxabs.hpp:44]: also other in the familly

File: include/eve/module/core/pedantic/impl/fma.hpp (hunk 5 above)
Comment (jtlap) [reply include/eve/module/core/pedantic/impl/fma.hpp:55]: done

File: include/eve/module/core/regular/rec.hpp (hunk 16 above)
Comment (jtlap) [reply include/eve/module/core/regular/rec.hpp:78]: suppressed

File: include/eve/module/math/constant/quarter.hpp (hunk 17 above)
Comment (jtlap) [reply include/eve/module/math/constant/quarter.hpp:65]: done

File: include/eve/module/math/regular/impl/cosh.hpp (hunk 18 above)
Comment (jtlap) [reply include/eve/module/math/regular/impl/cosh.hpp:34]: done

File: test/unit/module/math/cosh.cpp (hunk 20 above)
Comment (jtlap) [reply test/unit/module/math/cosh.cpp:33]: done

File: include/eve/module/core/pedantic/impl/is_flint.hpp (hunk 6 above)
Comment (jtlap) [reply include/eve/module/core/pedantic/impl/is_flint.hpp:28]: done

File: include/eve/module/core/regular/binarize.hpp (hunk 7 above)
Comment (jtlap) [reply include/eve/module/core/regular/binarize.hpp:56]: done

File: include/eve/module/core/regular/impl/three_fma.hpp (hunk 8 above)
Comment (jtlap) [reply include/eve/module/core/regular/impl/three_fma.hpp:10]: removed

File: include/eve/module/core/regular/impl/rec.hpp (hunk 9 above)
Comment (jtlap) [reply include/eve/module/core/regular/impl/rec.hpp:35]: done

File: include/eve/module/core/regular/quick_two_sub.hpp (hunk 12 above)
Comment (jtlap) [reply include/eve/module/core/regular/quick_two_sub.hpp:63]: suppressed

File: include/eve/module/core/regular/two_sqr.hpp (hunk 13 above)
Comment (jtlap) [reply include/eve/module/core/regular/two_sqr.hpp:61]: suppressed

File: include/eve/module/core/regular/two_sub.hpp (hunk 14 above)
Comment (jtlap) [reply include/eve/module/core/regular/two_sub.hpp:61]: suppressed

File: include/eve/module/core/saturated/impl/negmaxabs.hpp (hunk 15 above)
Comment (jtlap) [reply include/eve/module/core/saturated/impl/negmaxabs.hpp:45]: done
